toolchain/parse/typed_nodes.h

```diff
@@ -0,0 +1,781 @@
+  InterfaceIntroducerId introducer;
+  llvm::SmallVector<AnyModifier> modifiers;
+  AnyNameComponent name;
+  std::optional<ImplicitParamListId> implicit_params;
+  std::optional<TuplePatternId> params;
+};
+
+// `interface I;`
+using InterfaceDecl =
+    InterfaceSignature<NodeKind::InterfaceDecl, NodeCategory::Decl>;
+// `interface I {`
+using InterfaceDefinitionStart =
+    InterfaceSignature<NodeKind::InterfaceDefinitionStart, NodeCategory::None>;
+
+// `interface I { ... }`
+struct InterfaceDefinition {
+  static constexpr auto Kind =
+      NodeKind::InterfaceDefinition.Define(NodeCategory::Decl);
+  InterfaceDefinitionStartId signature;
+  llvm::SmallVector<AnyDecl> members;
+};
+
+// `impl`...`as` declarations and definitions
+
+// `impl`
+using ImplIntroducer = LeafNode<NodeKind::ImplIntroducer>;
+// `as`
+using ImplAs = LeafNode<NodeKind::ImplAs>;
+
+// `forall [...]`
+struct ImplForall {
+  static constexpr auto Kind = NodeKind::ImplForall.Define();
+  ImplicitParamListId params;
+};
+
+// `impl T as I`
+template <const NodeKind& KindT, NodeCategory Category>
+struct ImplSignature {
+  static constexpr auto Kind = KindT.Define(Category);
+  ImplIntroducerId introducer;
+  llvm::SmallVector<AnyModifier> modifiers;
+  std::optional<ImplForallId> forall;
+  std::optional<AnyExpr> type_expr;
+  ImplAsId as;
+  AnyExpr interface;
+};
+
+// `impl T as I;`
+using ImplDecl = ImplSignature<NodeKind::ImplDecl, NodeCategory::Decl>;
+// `impl T as I {`
+using ImplDefinitionStart =
+    ImplSignature<NodeKind::ImplDefinitionStart, NodeCategory::None>;
+
+// `impl T as I { ... }`
+struct ImplDefinition {
+  static constexpr auto Kind =
+      NodeKind::ImplDefinition.Define(NodeCategory::Decl);
+  ImplDefinitionStartId signature;
+  llvm::SmallVector<AnyDecl> members;
+};
+
+// Named constraint declarations and definitions
+
+// `constraint`
+using NamedConstraintIntroducer = LeafNode<NodeKind::NamedConstraintIntroducer>;
+
+// `constraint NC`
+template <const NodeKind& KindT, NodeCategory Category>
+struct NamedConstraintSignature {
+  static constexpr auto Kind = KindT.Define(Category);
+  NamedConstraintIntroducerId introducer;
+  llvm::SmallVector<AnyModifier> modifiers;
+  AnyNameComponent name;
+  std::optional<ImplicitParamListId> implicit_params;
+  std::optional<TuplePatternId> params;
+};
+
+// `constraint NC;`
+using NamedConstraintDecl =
+    NamedConstraintSignature<NodeKind::NamedConstraintDecl, NodeCategory::Decl>;
+// `constraint NC {`
+using NamedConstraintDefinitionStart =
+    NamedConstraintSignature<NodeKind::NamedConstraintDefinitionStart,
+                             NodeCategory::None>;
+
+// `constraint NC { ... }`
+struct NamedConstraintDefinition {
+  static constexpr auto Kind =
+      NodeKind::NamedConstraintDefinition.Define(NodeCategory::Decl);
+  NamedConstraintDefinitionStartId signature;
+  llvm::SmallVector<AnyDecl> members;
+};
+
+// Define `FooId` as the id for type `Foo`
```

Comment:
Probably this was originally my phrasing, but I think this is clearer the other way around:
```suggestion
// Define `Foo` as the node type for the ID type `FooId`.
```